fraim/core/llms/litellm.py

```diff
@@ -90,8 +100,85 @@ def with_tools(self, tools: Sequence[BaseTool], max_tool_iterations: Optional[in
             additional_model_params=self.additional_model_params,
             max_tool_iterations=max_tool_iterations,
             tools=tools,
+            max_retries=self.max_retries,
+            base_delay=self.base_delay,
+            max_delay=self.max_delay,
+        )
+
+    async def _run_once_with_retry(
```

Comment:
LiteLLM (supposedly) has built-in support for retries with expontential backoff: https://docs.litellm.ai/docs/completion/reliable_completions

Worth checking if that works as advertised.